piet-coregraphics/src/gradient.rs

```diff
@@ -0,0 +1,151 @@
+        }
+        .unwrap_or(Color::BLACK)
+    }
+
+    pub(crate) fn fill(&self, ctx: &mut CGContextRef) {
+        let context_ref: *mut u8 = ctx as *mut CGContextRef as *mut u8;
+        match self.piet_grad {
+            FixedGradient::Radial(FixedRadialGradient {
+                center,
+                origin_offset,
+                radius,
+                ..
+            }) => {
+                let start_center = to_cgpoint(center + origin_offset);
+                let center = to_cgpoint(center);
+                unsafe {
+                    CGContextDrawRadialGradient(
+                        context_ref,
+                        self.cg_grad.as_concrete_TypeRef(),
+                        start_center,
+                        0.0,
+                        center,
+                        radius as CGFloat,
+                        0,
+                    )
+                }
+            }
+            FixedGradient::Linear(FixedLinearGradient { start, end, .. }) => {
+                let start = to_cgpoint(start);
+                let end = to_cgpoint(end);
+                unsafe {
+                    CGContextDrawLinearGradient(
+                        context_ref,
+                        self.cg_grad.as_concrete_TypeRef(),
+                        start,
+                        end,
+                        0,
+                    )
+                }
+            }
+        }
+    }
+}
+
+fn new_cg_gradient(stops: &[GradientStop]) -> CGGradient {
+    unsafe {
+        //FIXME: is this expensive enough we should be reusing it?
+        let space = CGColorSpace::create_with_name(kCGColorSpaceSRGB).unwrap();
+        let space_ref: *const u8 = &*space as *const CGColorSpaceRef as *const u8;
+        let mut colors = Vec::<CGColor>::new();
+        let mut locations = Vec::<CGFloat>::new();
+        for GradientStop { pos, color } in stops {
+            let (r, g, b, a) = Color::as_rgba(&color);
+            let color = CGColorCreate(space_ref as *const u8, [r, g, b, a].as_ptr());
+            let color = CGColor::wrap_under_create_rule(color);
+            colors.push(color);
+            locations.push(*pos as CGFloat);
+        }
+
+        let colors = CFArray::from_CFTypes(&colors);
+        let gradient = CGGradientCreateWithColors(
+            space_ref as *const u8,
+            colors.as_concrete_TypeRef(),
+            locations.as_ptr(),
+        );
+
+        CGGradient::wrap_under_create_rule(gradient)
+    }
+}
+
+fn to_cgpoint(point: Point) -> CGPoint {
+    CGPoint::new(point.x as CGFloat, point.y as CGFloat)
+}
+
+#[link(name = "CoreGraphics", kind = "framework")]
+extern "C" {
+    fn CGGradientGetTypeID() -> CFTypeID;
+    fn CGGradientCreateWithColors(
+        space: *const u8,
+        colors: CFArrayRef,
+        locations: *const CGFloat,
+    ) -> CGGradientRef;
+    fn CGColorCreate(space: *const u8, components: *const CGFloat) -> SysCGColorRef;
```

Comment:
```suggestion
    #[allow(improper_ctypes)]
    fn CGColorCreate(space: *const CGColorSpaceRef, components: *const CGFloat) -> SysCGColorRef;
```

This is valid because `CGColorSpaceRef` is in foreign_types style, that is it represents the opaque object. The `#allow` is because foreign_types is missing `#[repr(C)]` on this decl, which is a problem with upstream; probably best to include a comment pointing that out.